storage/redis/datasource.go

```diff
@@ -146,91 +159,312 @@ func (rs *KeySpecificStream) Close() error {
+		}
+
+		tx.Abort() // We only read data above, no need to risk failing now.
+
+		for { // inner for is calling RunWorkerInternal
+			tx := rs.stateStorage.BeginTransaction().WithPrefix(rs.streamID.AsPrefix())
+
+			err := rs.RunWorkerInternal(ctx, tx)
+			if errors.Cause(err) == execution.ErrNewTransactionRequired {
+				tx.Abort()
+				continue
+			} else if waitableError := execution.GetErrWaitForChanges(err); waitableError != nil {
+				tx.Abort()
+				err = waitableError.ListenForChanges(ctx)
+				if err != nil {
+					log.Println("redis worker: couldn't listen for changes: ", err)
+				}
+				err = waitableError.Close()
+				if err != nil {
+					log.Println("redis worker: couldn't close storage changes subscription: ", err)
+				}
+				continue
+			} else if err == execution.ErrEndOfStream {
+				err = tx.Commit()
+				if err != nil {
+					log.Println("redis worker: couldn't commit transaction: ", err)
+					continue
+				}
+				return
+			} else if err != nil {
+				tx.Abort()
+				log.Printf("redis worker: error running redis read batch worker: %s, reinitializing from storage", err)
+				break
+			}
+
+			err = tx.Commit()
+			if err != nil {
+				log.Println("redis worker: couldn't commit transaction: ", err)
+				continue
+			}
+		}
+	}
+}
+
+var outputQueuePrefix = []byte("$output_queue$")
+
+func (rs *RecordStream) RunWorkerInternal(ctx context.Context, tx storage.StateTransaction) error {
+	outputQueue := execution.NewOutputQueue(tx.WithPrefix(outputQueuePrefix))
+
 	if rs.isDone {
-		return nil, execution.ErrEndOfStream
+		err := outputQueue.Push(ctx, &QueueElement{
+			Type: &QueueElement_Error{
+				Error: execution.ErrEndOfStream.Error(),
+			},
+		})
+		if err != nil {
+			return errors.Wrapf(err, "couldn't push csv EndOfStream to output record queue")
+		}
+
+		log.Println("csv worker: ErrEndOfStream pushed")
+		return execution.ErrEndOfStream
 	}
 
-	if rs.counter == len(rs.keys) {
-		rs.isDone = true
-		return nil, execution.ErrEndOfStream
+	var allKeysForThisBatch []string
+	var newCursor uint64
+	var err error
+
+	if rs.isEntireDatabaseStream {
+		// Performing scan from cursor -> extracting `rs.batchSize` new elements
+		allKeysForThisBatch, newCursor, err = rs.client.Scan(rs.cursor, "*", int64(rs.batchSize)).Result()
+		if err != nil {
+			return errors.Wrapf(err, "couldn't scan new batch for redis worker")
+		}
+	} else {
+		// Taking first `rs.batchSize` elements from all keys needed
+		if len(rs.keys) < rs.batchSize {
+			allKeysForThisBatch = rs.keys
+		} else {
+			allKeysForThisBatch = rs.keys[:rs.batchSize]
+		}
 	}
 
-	key := rs.keys[rs.counter]
-	rs.counter++
+	batch := make([]*execution.Record, 0)
+	for i := 0; i < len(allKeysForThisBatch); i++ {
+		key := allKeysForThisBatch[i]
+		recordValues, err := rs.client.HGetAll(key).Result()
+		if err != nil {
+			return errors.Wrapf(err, "could't get hash for key %s", key)
+		}
+
+		// We skip this record
+		if len(recordValues) == 0 {
```

Comment:
Can this happen? What does this mean? Is it same as not found or can the value be empty. If it can be empty, the just return a record with only the key.

Reply:
explained irl